### ISSUE TITLE
Id3 Genre Multiple Value Split Issue

### ISSUE THREAD

**Issue by DividedSE** [Zeugma440/atldotnet#108]
## The problem
Hope You remember me, I used to help with the playlist problems.
I finally decided to switch from taglib#(dead?) to atl.net 
I already faced an issue that i mentioned below. pls keep this thread open, I'll also post my problems here.
 string result = iGenre.Trim().Replace("\0", ""); // code from the ID3.cs
can you change the replace value from null to something like ; or / so I can split the multiple genre string.

## Environment

* ATL version (or git revision) that exhibits the issue: 3.22
* Last ATL version that did not exhibit the issue (if applicable): Unknown
* OS/version used to run ATL: Win 10 20H2

## Details

Genre value in TagScanner 6.19 :  Ambient; Country Music; Country Pop; Electronic; Jazz; Pop Music; Rock; Soft Rock (Split with ;)
Genre value in exiftool: Ambient/Country Music/Country Pop/Electronic/Jazz/Pop Music/Rock/Soft Rock (Split with /)
Genre value in ATL3.22: AmbientCountry MusicCountry PopElectronicJazzPop MusicRockSoft Rock (that i can't split because of null join)

## Code To Reproduce Issue [ Good To Have ]

return track.Genre.Split().ToList();    cant put any split value....

**Comment by Zeugma440** [comment]
Hi @DividedSE, long time no see :)

**For reading**
I'd rather use the `Settings.DisplayValueSeparator` character to separate values, because you have control over it.

**For writing**
I'd use the same convention as when writing multiple artists :
```csharp
theTrack.Artist = "artist1" + ATL.Settings.DisplayValueSeparator + "artist2" + ATL.Settings.DisplayValueSeparator + "artist3";
theTrack.Save();
```
***

Would that be fine to you ?

Also, could you send me a sample file so that I can make sure everything works as intended ?

**Comment by DividedSE** [comment]
So this is blank mp3 with no id3tag. I added three genre values using MusicBee Audio Player
I tried to get the values using this code:- 

internal static List<string> GetGenre(Track track)
        {
            return track.Genre.Split(Settings.DisplayValueSeparator).ToList(); //Its not separating here in debug its still 1 string
        }

Tags.Genre = GetID3.GetGenre(theTrack);

Console.WriteLine(string.Join("; ",tags.Genre.ToArray()));

My result is :- RockPopCountry
[test.zip](https://github.com/Zeugma440/atldotnet/files/6945500/test.zip)

**Comment by Zeugma440** [comment]
I just pushed a new commit that should do the trick : 29c8fbdd015d56f259c56798fe346c94ace7fd74

Are you able to test with a local project ?

***

Also : ` pls keep this thread open, I'll also post my problems here` => Please try to open a different issue for each problem you have in order to keep things tidy. I'm not a big fan of using one big github issue for various problems.

**Comment by DividedSE** [comment]
Well I tried but not everything works. 
The file I provided , i used musicbee to store the multiple genres i.e. three (Rock, Pop & Country)
after your fix this I used to get the values:- which works on guess
```
track.Genre.Split("; ").ToList(); //I need to guess the separator.

var array = data.Split('-'); // data is Rock-Pop-Country
            string writtenGenre = string.Join(Settings.DisplayValueSeparator, array); 
            string expectedGenre = writtenGenre.Replace(ATL.Settings.DisplayValueSeparator, ATL.Settings.InternalValueSeparator);
            theTrack.Genre = expectedGenre;
            return theTrack.Save();
```
Output on Musicbee should on Rock; Pop; Country & on exiftool should be Rock/Pop/Country
but when i saved some value with atl it saved it as a single string which not separately with other software like musicbee(Rock-Pop-Country) or exiftool(Rock╦╡Pop╦╡Country)

I will suggest you should just return the string which is stored on TCON frame which maybe a single genre value without any '/0' or multiple values with /0 without removing /0 this part. and make /0 a "settings.something" to let user replace it with suitable character for display. for save let user give the split char & replace it with /0 in a single string & put it in TCON frame like reverse fetch. Hope I clarify it.
If you use MusicBee or exifTool we will see what i meant to say.

**Comment by Zeugma440** [comment]
> But Writing the Parts didn't

As a client of the library, you should always use `DisplayValueSeparator`, which is in turn converted to the adequate separator (`\0` in case of ID3v2, but can be anything else if you're using other formats).

`InternalValueSeparator` is for ATL internal use only, as stated by the documentation.


> This /0 separation issue is also a problem of some Custom values like TMOO, TIPL, TMCL etc

Are you actually using those, or is it a theoretical issue ?

**Comment by DividedSE** [comment]
Ok. Got it Now. Using DisplayValueSeparator for both read & write & it works. Problem Solved.

> InternalValueSeparator is for ATL internal use only, as stated by the documentation.
Shouldn't it then be private only & not accessed by end-user?

> Are you actually using those, or is it a theoretical issue ?
Yup, I am using those to for both Read & Write like any other Tagger.

**Comment by Zeugma440** [comment]
> Shouldn't it then be private only & not accessed by end-user?
It should be `internal`, actually but you're completely right about that

> Yup, I am using those to for both Read & Write like any other Tagger.
Then I still have some extra work to make all of these behave the same way. I'll ping you when it's done !

**Comment by Zeugma440** [comment]
Btw you got me very curious about which app you're maintaining. Care to tell me more about that ? :)

**Comment by DividedSE** [comment]
Well, it's not for any app. Programming is my hobby. I built an audio player for my uni project, which I still tried to improve in my spare time :)

**Comment by DividedSE** [comment]
I Just Noticed an Issue... 
I was using MusicBrainz Picard to autotag the mp3's Metadata, which store multiple mood values as single string joined by \0 in TMOO ID3 Tag which was successfully read by both ATL & Musicbee & I Can separate them easily.
But after I used that mp3 in musicbee to edit some tag & save it & it changed the TMOO Id3 tag to generic TXXX/MOOD tag.
So, In ATL we have to option to retrieve additional fields 
1) //which only gets one MOOD tags not all Tags & other values missing
`var value = track.AdditionalFields.TryGetValue("MOOD", out var data) ? data : null`
2) looping //still giving one value & other values missing
```
foreach (KeyValuePair<string, string> field in theTrack.AdditionalFields)
{
 if(field.key=="MOOD")
tags+=field.value+"-";
}
```
I guess TXXX/customnametags have a issue reading values ?

**Comment by Zeugma440** [comment]
Thanks for the detailed report. It's a good illustration about the limits of the current implementation of ID3v2.

I'll keep it in mind when improving that part.

**Comment by DividedSE** [comment]
Thanks. As my Implementation of ATL is currently only limited to ID3V2 only which is 90% completly & waiting for the APIC fixes in other bug report. I have some minor feature request if you can implement.
1) TSO2(Album Artist Sort - string), TSOC(Composer Sort -string), TCMP(Compilation - stores 1 if song is a part of a compilation or remover the tag if not) is some Unoffical Id3 Tags used by some modern apps like itunes & Musicbrainz also currently recognised by ID3.Org. ATL  can read them from native fields but writes them in Txxx/... Instead of native ... Frame.
2) Implemention of reading file Access permision like "-rw-rw-rw" with (???)in Techninal Information with like last accessed, last modified, file created. I can do last 3 with File class but unable to get the permission part figure out.
3) Also An option to backup the file on track.save(); (i can implement in my code outside atl, just an idea if you want to implement in atl itself or not)
4)

**Comment by Zeugma440** [comment]
> I maybe asking too much. but everything is just a casual request. I liked how playlist thing completly 💯 done in the past & want to ATL be All in One solution for everyone to use from beginner to Pro.

I completely agree with that vision. Plus everything you've suggested has been relevant so far => please continue to do so, _but_ please also try to open distinct issues when your requests are too different from the initial issue 😅 

The last 3 you posted here really need to be each in their separate issue. Thanks~~ 😁

**Comment by DividedSE** [comment]
Well Ok Then Deleting that comment with separate issue.

**Comment by Zeugma440** [comment]
All ID3v2 fields that may contain multiple values are now handled by ATL, using `Settings.DisplayValueSeparator` as a separator at read time _and_ write time.

NB : The value separator used inside the updated file is determined according to the version of ID3v2 :
- `/` for ID3v2.2 and ID3v2.3
- `\0` for ID3v2.4

**Attention** You will have to update your code if you are manually separating values using `\0` or `/` when reading those fields. The separator to use is now `Settings.DisplayValueSeparator`.

**Comment by DividedSE** [comment]
I forgot to reply, yes i updated my code(replaced \0 with Settings.DisplayValueUpdater) & it works normally without an issue(only tested v2.4).
I will test 2.3 & get u back.
P.s. my reply might get late due to some issues . Due to job i can now only have time for this on sundays.

**Comment by Zeugma440** [comment]
The fix has been published on today's v3.25
